contracts/Governance/TimelockV8.sol

```diff
@@ -0,0 +1,226 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.13;
+
+/**
+ * @title TimelockV8
+ * @author Venus
+ * @notice The Timelock contract using solidity V8.
+ * This contract also differs from the original timelock because it has a virtual function to get minimum delays
+ * and allow test deployments to override the value.
+ */
+contract TimelockV8 {
+    /// @notice Event emitted when a new admin is accepted
+    event NewAdmin(address indexed newAdmin);
+
+    /// @notice Event emitted when a new admin is proposed
+    event NewPendingAdmin(address indexed newPendingAdmin);
+
+    /// @notice Event emitted when a new admin is proposed
+    event NewDelay(uint256 indexed newDelay);
+
+    /// @notice Event emitted when a proposal transaction has been cancelled
+    event CancelTransaction(
+        bytes32 indexed txHash,
+        address indexed target,
+        uint256 value,
+        string signature,
+        bytes data,
+        uint256 eta
+    );
+
+    /// @notice Event emitted when a proposal transaction has been executed
+    event ExecuteTransaction(
+        bytes32 indexed txHash,
+        address indexed target,
+        uint256 value,
+        string signature,
+        bytes data,
+        uint256 eta
+    );
+
+    /// @notice Event emitted when a proposal transaction has been queued
+    event QueueTransaction(
+        bytes32 indexed txHash,
+        address indexed target,
+        uint256 value,
+        string signature,
+        bytes data,
+        uint256 eta
+    );
+
+    /// @notice Required period to execute a proposal transaction
+    uint256 private constant GRACE_PERIOD = 14 days;
+
+    /// @notice Minimum amount of time a proposal transaction must be queued
+    uint256 private constant MINIMUM_DELAY = 1 hours;
+
+    /// @notice Maximum amount of time a proposal transaction must be queued
+    uint256 private constant MAXIMUM_DELAY = 30 days;
```

Comment:
Nice trick with overrides! What bothers me here, though, is that we'll have two similar entities per parameter (a constant and a getter) that may potentially return different values, and it harms readability a bit. Maybe rename the constants to sth like `DEFAULT_ GRACE_PERIOD` to make sure it won't be blindly used instead of getGracePeriod() in some future reincarnation of the timelock?

An alternative solution could use immutables instead of / in addition to inheritance here, setting values of these immutables in the constructor. This could look like:

```
contract TimelockV8 {
    uint256 public immutable gracePeriod = 14 days;
    uint256 public immutable minimumDelay = 1 hours;
    uint256 public immutable maximumDelay = 30 days;
    
    ...
    
    constructor(
        address admin_,
        uint256 delay_,
        uint256 gracePeriod_,
        uint256 minimumDelay_,
        uint256 maximumDelay_
    ) public { ... }

    ...
}
```

The downside is, ofc, having 5 params in the constructor. Inheritance can solve this, but with two extra contracts:

```
abstract contract BaseTimelockV8 {
    /* same as above */
}

contract MainnetTimelockV8 is BaseTimelockV8 {
    constructor(address admin_, uint256 delay_)
        BaseTimelockV8(admin_, delay_, 14 days, 1 hours, 30 days)
    { }
}

contract TestnetTimelockV8 is BaseTimelockV8 {
    constructor(address admin_, uint256 delay_)
        BaseTimelockV8(admin_, delay_, 1, 1, 1 hours)
    { }
}
```

I don't have any preference here, probably the `DEFAULT_*` thing is easiest

Reply:
I thought about this which is why I made them private. Naming them default would be clearer. 

We discussed adding the defaults to the constructor but that means it isn't really worth checking the values in the constructor so we could remove those checks. If we remove those checks we inadvertently solve the problem and don't need to add defaults to the constructor.

With this strategy we end up keeping the checks in the constructor and still have the flexibility to deploy with different values to test networks